File: src/java/magma/Path_.java
package magma;

import magma.collect.Set_;
import magma.collect.stream.Stream;

import java.io.IOException;
import java.nio.file.Files;

public interface Path_ {
    default boolean exists() {
        return Files.exists((Paths.toNative(this)));
    }

    Set_<Path_> walk() throws IOException;

    void writeString(String output) throws IOException;

    void createAsDirectories() throws IOException;

    Path_ resolve(String child);

    Stream<String> stream();

    Path_ relativize(Path_ child);

    boolean isRegularFile();

    String asString();

    Path_ getParent();

    Path_ getFileName();

    String readString() throws IOException;
}

File: src/windows/magma/api/io/Path.c
#include "./Path.h"
export interface Path {
	mut asString(): &[I8];
	mut writeString(mut output: &[I8]): Option<IOError>;
	mut readString(): Result<&[I8], IOError>;
	mut walk(): Result<List<Path>, IOError>;
	mut findFileName(): &[I8];
	mut endsWith(mut suffix: &[I8]): Bool;
	mut relativize(mut source: Path): Path;
	mut getParent(): Path;
	mut query(): Query<&[I8]>;
	mut resolveChildSegments(mut children: List<&[I8]>): Path;
	mut resolveChild(mut name: &[I8]): Path;
	mut exists(): Bool;
	mut createDirectories(): Option<IOError>;
}
